inception-workload-dynamic/src/main/java/de/tudarmstadt/ukp/inception/workload/dynamic/extensions/DynamicAnnotatorWorkflowActionBarItemGroup.java

```diff
@@ -0,0 +1,141 @@
+        String aId, AnnotationPageBase aPage, EntityManager aEntityManager)
+    {
+        super(aId);
+
+        //Same as for the default
+        page = aPage;
+        entityManager = aEntityManager;
+
+        add(finishDocumentDialog = new ConfirmationDialog("finishDocumentDialog",
+            new StringResourceModel("FinishDocumentDialog.title", this, null),
+            new StringResourceModel("FinishDocumentDialog.text", this, null)));
+
+        add(finishDocumentLink = new LambdaAjaxLink("showFinishDocumentDialog",
+            this::actionFinishDocument));
+        finishDocumentLink.setOutputMarkupId(true);
+        finishDocumentLink.add(enabledWhen(() -> page.isEditable()));
+        finishDocumentLink.add(new Label("state")
+            .add(new CssClassNameModifier(LambdaModel.of(this::getStateClass))));
+    }
+
+    protected AnnotationPageBase getAnnotationPage()
+    {
+        return page;
+    }
+
+    public String getStateClass()
+    {
+        return FontAwesome5IconType.check_circle_r.cssClassName();
+    }
+
+    protected void actionFinishDocument(AjaxRequestTarget aTarget)
+    {
+        finishDocumentDialog.setConfirmAction((_target) ->
+        {
+            page.actionValidateDocument(_target, page.getEditorCas());
+
+            AnnotatorState state = page.getModelObject();
+            User user = state.getUser();
+            Project project = state.getProject();
+            SourceDocument document = state.getDocument();
+
+            AnnotationDocument annotationDocument = documentService
+                .getAnnotationDocument(document, user);
+
+            documentService.transitionAnnotationDocumentState(annotationDocument,
+                ANNOTATION_IN_PROGRESS_TO_ANNOTATION_FINISHED);
+
+            //Go through all documents in a random order and check if there is a Annotation document
+            //with the state NEW
+            String query =  "FROM SourceDocument " +
+                            "WHERE project = :project " +
+                            "ORDER BY rand()";
+            for (SourceDocument doc: entityManager.createQuery(query,SourceDocument.class)
+                .setParameter("project", project).getResultList()) {
+                //Check if it even exists or is state NEW
+                if (documentService.listAnnotatableDocuments(project,user).get(doc) == null ||
+                    documentService.listAnnotatableDocuments(project,user).get(doc).
+                        getState().equals(NEW)) {
+                    getAnnotationPage().getModelObject().setDocument(doc, documentService.
+                        listSourceDocuments(project));
+                    //This document had the state NEW, load it
+                    Optional<AjaxRequestTarget> target = RequestCycle.get().
+                        find(AjaxRequestTarget.class);
+                    getAnnotationPage().actionLoadDocument(target.orElse(null));
```

Comment:
You have to use the `_target` passed to the confirm action (not `target`) within the confirm action - when you do that, there should be no need to fetch it from the request cycle.